File: ghidra_scripts/decompile_all.java
// Decompile all functions from a binary using Ghidra Headless.
// @author slopwise
// @category slopwise

import ghidra.app.script.GhidraScript;
import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompileResults;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionIterator;
import ghidra.program.model.listing.Program;
import java.io.PrintWriter;
import java.io.File;

public class decompile_all extends GhidraScript {
    @Override
    protected void run() throws Exception {
        String[] args = getScriptArgs();
        if (args.length < 1) {
            println("Usage: decompile_all.java <output_json_path>");
            return;
        }
        
        File outFile = new File(args[0]);
        PrintWriter writer = new PrintWriter(outFile, "UTF-8");

        Program program = getCurrentProgram();
        if (program == null) {
            writer.println("[]");
            writer.close();
            return;
        }

        DecompInterface decompiler = new DecompInterface();
        decompiler.openProgram(program);
        
        FunctionIterator functions = program.getFunctionManager().getFunctions(true);
        
        writer.println("[");
        boolean first = true;
        for (Function func : functions) {
            if (func.isExternal() || func.isThunk()) {
                continue;
            }
            
            DecompileResults decompRes = decompiler.decompileFunction(func, 60, monitor);
            String decompiledCode = "/* Decompilation failed */";
            if (decompRes != null && decompRes.decompileCompleted()) {
                decompiledCode = decompRes.getDecompiledFunction().getC();
            }
            
            if (!first) {
                writer.println(",");
            }
            first = false;
            
            String name = escapeJson(func.getName());
            String signature = escapeJson(func.getPrototypeString(true, true));
            String address = escapeJson(func.getEntryPoint().toString());
            String code = escapeJson(decompiledCode);
            
            writer.println("  {");
            writer.println("    \"name\": \"" + name + "\",");
            writer.println("    \"signature\": \"" + signature + "\",");
            writer.println("    \"address\": \"" + address + "\",");
            writer.println("    \"decompiled\": \"" + code + "\"");
            writer.print("  }");
        }
        writer.println("\n]");
        writer.close();
    }

    private String escapeJson(String input) {
        if (input == null) return "";
        return input.replace("\\", "\\\\")
                    .replace("\"", "\\\"")
                    .replace("\b", "\\b")
                    .replace("\f", "\\f")
                    .replace("\n", "\\n")
                    .replace("\r", "\\r")
                    .replace("\t", "\\t");
    }
}

File: ghidra_scripts/decompile_all.py
# Decompile all functions from a binary using Ghidra Headless.
# @author slopwise
# @category slopwise
# @keybinding 
# @menupath 
# @toolbar 

import json
import sys
from ghidra.app.decompiler import DecompInterface
from ghidra.util.task import ConsoleTaskMonitor

def run():
    program = getCurrentProgram()
    if not program:
        print("Error: No program loaded")
        return

    decompiler = DecompInterface()
    decompiler.openProgram(program)
    
    function_manager = program.getFunctionManager()
    functions = function_manager.getFunctions(True) # True = forward
    
    results = []
    monitor = ConsoleTaskMonitor()
    
    for func in functions:
        # Skip external/thunk functions if they don't have code
        if func.isExternal() or func.isThunk():
            continue
            
        decomp_res = decompiler.decompileFunction(func, 60, monitor)
        if decomp_res and decomp_res.decompileCompleted():
            results.append({
                "name": func.getName(),
                "signature": func.getPrototypeString(True, True),
                "decompiled": decomp_res.getDecompiledFunction().getC(),
                "address": func.getEntryPoint().toString()
            })
        else:
            # Fallback for functions that fail to decompile
            results.append({
                "name": func.getName(),
                "signature": func.getPrototypeString(True, True),
                "decompiled": "/* Decompilation failed */",
                "address": func.getEntryPoint().toString(),
                "error": str(decomp_res.getErrorMessage()) if decomp_res else "Unknown error"
            })
            
    # Print the JSON to a marker so we can extract it from the logs
    print("---SLOPWISE-START---")
    print(json.dumps(results, indent=2))
    print("---SLOPWISE-END---")

if __name__ == "__main__":
    run()
